lib/ws/ws.js

```diff
@@ -9,13 +9,86 @@ var factory = require('./factory');
 var authMethod = 'auth';
 
 /**
+ * race takes an array of promises and is resolved with the first promise
+ * that resolves, or rejects. After the first resolve or rejection, all
+ * future resolutions and rejections will be discarded.
+ * @param {Promise[]} promises
+ * @return {Promise}
+ */
+function race(promises) {
+    return new BeamSocket.Promise(function (_resolve, _reject) {
+        var done = false;
+        function guard(fn) {
+            return function () {
+                if (!done) {
+                    done = true;
+                    fn.apply(null, arguments);
+                }
+            };
+        }
+
+        var resolve = guard(_resolve);
+        var reject = guard(_reject);
+
+        for (var i = 0; i < promises.length; i++) {
+            promises[i].then(resolve).catch(reject);
+        }
+    });
+}
+
+/**
+ * Return a promise which is rejected with a TimeoutError after the
+ * provided delay.
+ * @param  {Number} delay
+ * @return {Promise}
+ */
+function timeout(delay) {
+    return new BeamSocket.Promise(function (resolve, reject) {
+        setTimeout(function () {
+            reject(new TimeoutError());
+        }, delay);
+    });
+}
+
+/**
+ * Assigns enumerable properties of the src onto the obj, returning src.
+ * Semi-polyfill for Object.assign.
+ * @param  {Object} obj
+ * @param  {Object} src
+ * @return {Object}
+ */
+function assign(src, obj) {
```

Comment:
lodash is a dependency of this module, could we not use https://lodash.com/docs#assign